Fix slf4j LinkageErrors for external plugins

eyalkoren opened the pull request:
## What does this PR do?
<!--
Replace this comment with a description of what's being changed by this PR.
Please explain the WHAT: A clear and concise description of what (patterns used, algorithms implemented, design architecture, message processing, etc.)
Can be as simple as Fixes #123 if there's a corresponding issue.
-->
Fixes #2350 

This PR touches several places related to external plugins:
- Class loading (@felixbarny please review this part)
     - The CL hierarchy for external plugin is currently not what we intended for the API classes. If we load the API classes by the IndyPluginCL, then the API plugin classes are loaded by another IndyPluginCL, one of which parents is the agent CL and the other is the external plugin IndyPluginCL. So I made sure those are loaded by the `ExternalPluginClassLoader`. 
     - If slf4j classes are packaged into the plugin, I made sure to exclude them as well from the external plugin IndyPluginCL, because it is child-first, which means it will cause clashes with the slf4j that is available in the agent CL. It means that the `ExternalPluginClassLoader` will still have visibility to slf4j, but those cannot effectively be loaded, so the plugin slf4j usages will use the agent slf4j.
     - Our API plugin advice uses the instrumented class's (the API class) CL as the "initiating CL" (a term with a different meaning in the Java spec BTW 🙂 ). When the API is added to an application as a dependency, this makes perfect sense. But when it is loaded by an external plugin CL, this doesn't make sense. I changed it currently so that `TraceContext` will just ignore attempts to set any type of agent CL as the initiating CL. This is enough to cause the usage of the context CL as a fallback, which will probably be sufficient in many cases. #2369 contains a discussion with multiple proposals of how to improve that. Let's discuss this offline.
- External plugin test apps (@jackshirazi please review this part)
     - Added ability to create an external-plugin-based transaction (was only span creation before)
     - Added log correlation by default
     - Added slf4j to the external plugin. This with the two above reproduced #2350 
     - Added slf4j to the external plugin target class (just to make sure we don't break stuff this way)
     - Added Jakarta version for the external plugin test app
     - Fixed and added tests for automatic outcome setting for API spans/transactions when capturing Exceptions

## Checklist
<!--
You only have to fill one category of checkboxes, depending on the type of the PR.
Please DO NOT remove any item, ~~striking through~~ those that do not apply.
Just ignore the checkboxes of categories that don't apply.
-->
- [x] This is a bugfix
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that would fail without this fix

apmmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-2376/detail/PR-2376/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-2376/detail/PR-2376/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-2376/detail/PR-2376/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-2376/detail/PR-2376/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-java_2376.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2021-12-29T07:57:03.645Z&rangeTo=2021-12-29T08:17:03.645Z&transactionName=apm-agent-java%2Fapm-agent-java-mbp%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=2cb1045f5e803b9ee7d5656c68a52243&transactionId=a9b565b4a0df09d4)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2021-12-29T08:07:03.645+0000


* Duration: 66 min 3 sec


* Commit: 3473586c79249381dae041ea7b4572624513fe26



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2642  |
| Skipped      | 21 |
| Total        | 2663   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `run benchmark tests` : Run the benchmark test.


- `run compatibility tests` : Run the JDK Compatibility test.


- `run integration tests` : Run the APM-ITs.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

eyalkoren commented:
@jackshirazi just proposed that we get rid of both log4j and slf4j by taking only the bare minimum from whichever logging framework we choose and repackage it. It sounds like it could save us ton of trouble

felixbarny commented:
> @jackshirazi just proposed that we get rid of both log4j and slf4j by taking only the bare minimum from whichever logging framework we choose and repackage it. It sounds like it could save us ton of trouble

This may make it harder to use ecs-logging-java. If feasible, I'd like to keep us using the standard slf4j and log4j libraries. However, we can definitely think about stripping unused classes if possible. For example, `org.apache.logging.log4j.core.appender.<db|mom|nosql>`.

> If slf4j classes are packaged into the plugin

We may also just fail to load plugins that include classes that shouldn't be in there (slf4j, Byte Buddy, the Plugin SDK) by throwing an exception. This seems a bit safer as no loading them as there may be binary incompatibilities. But probably not a big issue.

> Our API plugin advice uses the instrumented class's (the API class) CL as the "initiating CL" 

Could we detect if the initiating class loader is the indy plugin CL and programmatically access the target CL?

eyalkoren commented:
> We may also just fail to load plugins that include classes that shouldn't be in there (slf4j, Byte Buddy, the Plugin SDK) by throwing an exception. This seems a bit safer as no loading them as there may be binary incompatibilities. But probably not a big issue.

What if we just remove slf4j and use our own facade over log4j? If we add that through the SDK, external plugins can log into agent logs as well

> Could we detect if the initiating class loader is the indy plugin CL and programmatically access the target CL?

Yes, I have this working. The problem is that for external plugin, the target CL is not necessarily any better, it could be just the instrumented library CL that probably doesn't have access to the logging library.

eyalkoren commented on apm-agent-core/src/main/java/co/elastic/apm/agent/bci/IndyBootstrap.java:
```diff
@@ -362,10 +363,22 @@ public static ConstantCallSite bootstrap(MethodHandles.Lookup lookup,
+                    if (// API classes have no dependencies and don't need to be loaded by an IndyPluginCL
+                        !(externalPluginClass.startsWith("co.elastic.apm.api")) &&
+                        !(externalPluginClass.startsWith("co.elastic.apm.opentracing")) &&
+                        // slf4j classes, if used by the plugin, should be loaded by the agent Class Loader and not the
+                        // IndyPluginCL, which is child-first, as this would cause LinkageErrors due to clash with the agent's slf4j
+                        !(externalPluginClass.startsWith("org.slf4j"))
```
We don't need that if we merge #2390

jackshirazi commented on apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java:
```diff
@@ -712,7 +713,7 @@ public int hashCode() {
 
     void setApplicationClassLoader(@Nullable ClassLoader classLoader) {
-        if (classLoader != null) {
+        if (classLoader != null && !ClassLoaderUtils.isAgentClassLoader(classLoader)) {
```
`classLoader != null` test unnecessary now

eyalkoren replied:
It _is_ necessary - if `classLoader == null`, then `!ClassLoaderUtils.isAgentClassLoader(classLoader)` would evaluate to `true` and enter the `if` block, which is not what we want.

felixbarny replied:
[minor] I think the expression would be easier to grasp if inverted and returning

```java
if (classLoader == null || ClassLoaderUtils.isAgentClassLoader(classLoader)) {
    return;
}
``` 

Or even


```java
if (ClassLoaderUtils.isBootstrapClassLoader(classLoader) || ClassLoaderUtils.isAgentClassLoader(classLoader)) {
    return;
}
```

eyalkoren replied:
OK

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java:
```diff
@@ -390,14 +390,17 @@ public boolean isExit() {
-    public void captureException(long epochMicros, Throwable t) {
-        tracer.captureAndReportException(epochMicros, t, this);
+    @Nullable
+    public String captureException(long epochMicros, @Nullable Throwable t) {
```
It should either return `T` or be named `captureExceptionAndGetErrorId`.

jackshirazi commented on integration-tests/application-server-integration-tests/src/test/java/co/elastic/apm/servlet/AbstractTomcatIT.java:
```diff
@@ -33,7 +33,7 @@ public AbstractTomcatIT(final String tomcatVersion) {
     @Override
     protected void enableDebugging(GenericContainer<?> servletContainer) {
-        servletContainer.withEnv("CATALINA_OPTS", "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005");
+        servletContainer.withEnv("CATALINA_OPTS", "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:5005");
```
is `suspend=y` right?

eyalkoren replied:
Since debugging is disabled by default (through `co.elastic.apm.servlet.AbstractServletContainerIntegrationTest#ENABLE_DEBUGGING`), this is not super crucial. I think that I use `suspend=y` more often than not, but let's keep `n` as the default.

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/bci/IndyBootstrap.java:
```diff
@@ -362,10 +363,22 @@ public static ConstantCallSite bootstrap(MethodHandles.Lookup lookup,
             List<String> pluginClasses = new ArrayList<>();
             if (instrumentationClassLoader instanceof ExternalPluginClassLoader) {
-                pluginClasses.addAll(((ExternalPluginClassLoader) instrumentationClassLoader).getClassNames());
+                List<String> externalPluginClasses = ((ExternalPluginClassLoader) instrumentationClassLoader).getClassNames();
```
Possibly outside the scope of this PR but I think we should add more validation rules in the constructor of `ExternalPluginClassLoader` to make sure external plugins don't contain slf4j and Byte Buddy, similar to how we validate that the Plugin SDK isn't packaged in the plugin.

https://github.com/elastic/apm-agent-java/blob/e756114a054faf650e92e62b1793b0a1aa996e5a/apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java#L46-L48

While looking at it, the `%s` placeholder is unresolved.

eyalkoren replied:
Yes, let's handle separately (except from the Exception message String format).
With #2390 would be allowed, but all dependencies we package within the agent (not only Byte Buddy) may indeed cause `LinkageError` if used within the external plugin AND available to the CL that loads the instrumented classes.

eyalkoren replied:
I was primarily thinking about `IndyPluginClassLoader` created for external plugins' instrumentation, which may cause this specific kind of `LinkageError` - "loader constraint violation" (because of the child-first nature of this CL). 
However, since the agent CL is a direct parent of the `ExternalPluginClassLoader`, packaging classes that are available within the agent may either pass quietly, or cause other types of `LinkageError` (e.g. when using mismatching versions).
So I take this back - it is within the context of this PR to disallow external plugins to package libraries that are packaged in the agent.

felixbarny reviewed (approved):
We should probably merge #2390 first.

eyalkoren commented:
I added a verification mechanism to ensure that external plugins that contain agent dependencies are rejected with a proper log message.
I also added a test to ensure that the dependencies list is kept up to date

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java:
```diff
@@ -55,7 +58,27 @@ public ExternalPluginClassLoader(File pluginJar, ClassLoader agentClassLoader) t
                 if (jarEntry.getName().endsWith(".class")) {
-                    tempClassNames.add(jarEntry.getName().replace('/', '.').substring(0, jarEntry.getName().length() - 6));
+                    String fqcn = jarEntry.getName().replace('/', '.').substring(0, jarEntry.getName().length() - 6);
+                    if (fqcn.startsWith("org.slf4j") || fqcn.startsWith("org.apache.logging")) {
```
These are also in the list of agent dependencies so this case is not strictly needed. But I suppose you still opted for it because of the better error message, right?

eyalkoren replied:
Right

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java:
```diff
@@ -55,7 +58,27 @@ public ExternalPluginClassLoader(File pluginJar, ClassLoader agentClassLoader) t
+                            LoggerFactory.class.getName()
+                        ));
+                    }
+                    for (String agentDependencyPackage : AgentInfo.getAgentDependencyPackages()) {
```
Could we make the `ExternalPluginClassLoader` child-first and allow plugins to have a dependency on `HdrHistrogram`, for example?

eyalkoren replied:
It will create further complications...
We create an `IndyPluginClassLoader` for each external plugin, which can load most of the plugin classes. If you leave it as is (i.e. include these libraries within the IndyPlugin CL classpath), then you may get `LinkageError` if the `ExternalPluginClassLoader ` also loads these classes (e.g. because the instrumentation class uses it), because it is in the plugin CL delegation hierarchy. If you exclude these classes from the ones loaded by the Indy Plugin CL, then by default it would delegate loading requests of them to the agent CL (because this is how the `DiscriminatingMultiParentClassLoader` works), which may lead to other `LinkageError`
